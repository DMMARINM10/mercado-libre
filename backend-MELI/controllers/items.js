const { request, response } = require("express");
const { numberSeparation, arrayCategories } = require("../helpers/utils");

const searchUrl = process.env.MERCADO_LIBRE_SEARCH_RESULTS_ENDPOINT;
const itemUrl = process.env.MERCADO_LIBRE_ITEM_ENDPOINT;
const categoryUrl = process.env.MERCADO_LIBRE_CATEGORY_ENDPOINT;

const author = {
    name: "Danna",
    lastname: "Marin",
}

const getSearchResults = async (req = request, res = response) => {
    const { q: query } = req.query;

    if (!query) {
        return res.status(400).json({
            message: 'Search query should not be empty',
        });
    }

    try {
        const data = await fetch(`${searchUrl}?q=${query}&limit=4`);
        const jsonData = await data.json();
        const categories = arrayCategories(jsonData);
        const items = jsonData.results.map((item) => {
            const { id, title, sale_price, thumbnail: picture, shipping, condition, seller } = item;
            const { amount, currency_id: currency } = sale_price;
            const { free_shipping } = shipping;
            const { nickname } = seller;
            const { integer, decimals } = numberSeparation(amount);
            return {
                id,
                title,
                price: {
                    currency,
                    amount: integer,
                    decimals,
                },
                picture,
                free_shipping,
                condition,
                seller: nickname,
            }
        })
        res.json({
            author,
            categories,
            items,
        })
    } catch (error) {
        console.error(error);
        res.status(500).json({
            msg: 'Internal Server Error',
            error
        })
    }
};

const getItem = (req = request, res = response) => {
    const { id } = req.params;
    const productUrl = `${itemUrl}/${id}`;

    const fetchItem = fetch(`${productUrl}`).then(response => response.json());
    const fetchItemDescription = fetch(`${productUrl}/description`).then(response => response.json());
    
    Promise.all([fetchItem, fetchItemDescription])
    .then(([dataItem, dataDescription]) => {
        if (dataItem.error || dataDescription.error) {
            return res.status(404).json({
                message: 'Resource not found'
            });
        }
        const { id,
            title,
            price,
            currency_id: currency,
            pictures,
            condition,
            shipping,
            initial_quantity: sold_quantity,
            category_id,
        } = dataItem;
        const { plain_text: description } = dataDescription;
        const { integer, decimals } = numberSeparation(price);
        const picture = pictures[0].secure_url;
        const { free_shipping } = shipping;

        fetch(`${categoryUrl}/${category_id}`)
        .then((response) => response.json())
        .then(response => {
                if (response.error) {
                    return res.status(404).json({
                        message: 'Resource not found'
                    });
                }
                const { path_from_root } = response;
                const categories = path_from_root.map((path) => path.name)
                res.json({
                    author,
                    item: {
                        id,
                        title,
                        price: {
                            currency,
                            amount: integer,
                            decimals: decimals,
                        },
                        picture,
                        condition,
                        free_shipping,
                        sold_quantity,
                        description,
                        categories,
                    }
                })
            })
            .catch(error => {
                console.error(error);
                res.status(500).json({
                    msg: 'Internal Server Error',
                    error
                })
            });
        })
        .catch(error => {
            console.error(error);
            res.status(500).json({
                msg: 'Internal Server Error',
                error
            })
        });

};

module.exports = {
    getItem,
    getSearchResults,
}