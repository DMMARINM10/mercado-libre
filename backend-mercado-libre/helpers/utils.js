const numberSeparation = (number) => {
    const integer = Math.floor(number);
    const decimals = Number((number - integer).toFixed(2));
    return {
        integer,
        decimals,
    }
}

const arrayCategories = (data) => {
    const availableFilters = data.available_filters.find((filter) => filter.id === 'category');
    if(availableFilters) {
        const categories = availableFilters?.values?.sort((a, b) => b.results - a.results).map((value) => value.name);
        return categories;
    }
    const filter = data?.filters?.find((filter) => filter?.id === 'category');
    const categories = filter?.values[0]?.path_from_root?.map((cat) => cat.name)
    return categories;
}

module.exports = {
    numberSeparation,
    arrayCategories,
}