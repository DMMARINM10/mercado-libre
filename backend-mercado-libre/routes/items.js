const { Router } = require('express');
const { getItem, getSearchResults } = require('../controllers/items.js');

const router = Router();

router.get('/:id', getItem);

router.get('/', getSearchResults);

module.exports = router;