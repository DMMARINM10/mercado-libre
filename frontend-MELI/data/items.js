const ITEM_CONDITION = {
    new: 'Nuevo',
    used: 'Usado',
}

export const itemCondition = (condition) => {
    return ITEM_CONDITION[condition] || '';
}