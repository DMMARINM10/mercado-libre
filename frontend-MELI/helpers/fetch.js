export const fetcher = (url) => {
    return fetch(url)
    .then((res) => {
        if(!res.ok) {
            throw new Error()
        }
        return res.json()
    });
}

export const getItem = async (id) => {
    const url = `${process.env.NEXT_PUBLIC_API_URL}/${id}`
    try {
        const resp = await fetch(url)
        if(!resp.ok) {
            return {
                error: true,
                data: null,
            }
        }
        const data = await resp.json()
        return {
            data,
            error: false
        }
    } catch (error) {
        console.error(error)
        return {
            error: true,
            data: null,
        }
    }
}