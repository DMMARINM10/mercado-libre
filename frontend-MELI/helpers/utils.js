export const formatNumber = (number) => {
    return number.toString().replace(/\B(?=(\d{3})+(?!\d))/g, ".");
}

export const decimalFormat = (number) => {
    const decimalString = number?.toString()
    const result = decimalString?.includes('.') 
        ? decimalString.substring(2, decimalString.length) 
        : '00'
    return result?.length !== 2 ? result + '0' : result
}