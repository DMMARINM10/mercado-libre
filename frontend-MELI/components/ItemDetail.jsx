import Image from "next/image"
import { itemCondition } from "@/data/items"
import { formatNumber } from "@/helpers/utils"
import styles from '@/styles/components/itemDetail.module.scss'

const ItemDetail = ({ id, picture, description, condition, soldQuantity, title, price }) => {
  return (
    <>
      <div className={styles['item-detail-left']}>
        <Image src={picture} alt={`picture-item-${id}`} width={680} height={680} />
        <div className={styles['item-description-container']}>
            <h2>Descripci√≥n del producto</h2>
            <p>{description}</p>
        </div>
      </div>
      <div className={styles['item-detail-right']}>
        <p className={styles['item-detail-condition']}>{itemCondition(condition)} - {soldQuantity} vendidos</p>
        <div className={styles['item-detail-metadata']}>
          <h1>{title}</h1>
          <p>$ {formatNumber(price)}</p>
          <button>Comprar</button>
        </div>
      </div>
    </>
  )
}

export default ItemDetail
