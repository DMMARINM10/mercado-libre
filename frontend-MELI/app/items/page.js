'use client'

import useSWR from "swr"
import { useSearchParams } from "next/navigation"
import Breadcrumb from "@/components/Breadcrumbs"
import Card from "@/components/Card"
import Loader from "@/components/Loader"
import { fetcher } from "@/helpers/fetch"
import Error from "@/components/Error"
import { Suspense } from "react"


const Items = () => {
    const searchParams = useSearchParams()
    const search = searchParams.get('search')
    const { data, error, isLoading } = useSWR(
      search ? `${process.env.NEXT_PUBLIC_API_URL}?q=${search}` : null,
      fetcher
    );
    
    const categories = data?.categories
    const breadcrumb = categories?.slice(0, 4)
    const items = data?.items

    if(!search) return <></>
    
    if(error) return <Error />

    if(isLoading) return <Loader />
    
    return (
      <>
      {
        data &&
        <>
          <Breadcrumb breadcrumb={breadcrumb} />
          <section className="container">
          {
            items?.map((item) => {
              const { id, picture, price, title, seller, free_shipping: freeShipping } = item
              const { amount } = price
              return (
                <Card 
                 key={id}
                 id={id}
                 picture={picture}
                 amount={amount}
                 freeShipping={freeShipping}
                 title={title}
                 seller={seller}
                />
              )
            })
          }
          </section>
        </>
      }
      </>
    );
}

export default Items;
