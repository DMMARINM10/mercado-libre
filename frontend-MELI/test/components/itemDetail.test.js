import React from 'react';
import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import ItemDetail from '@/components/ItemDetail';
import { itemCondition } from '@/data/items';
import { formatNumber } from '@/helpers/utils';

// Mock next/image
jest.mock('next/image', () => {
  return ({ src, alt, ...props }) => <img src={src} alt={alt} {...props} />;
});

describe('ItemDetail component', () => {
  const props = {
    id: '1',
    picture: 'http://random-url.jpg',
    description: 'Test description',
    condition: 'new',
    soldQuantity: 10,
    title: 'Test title',
    price: 1000
  };

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should render the item detail with correct data', () => {
    render(<ItemDetail {...props} />);

    const image = screen.getByAltText(`picture-item-${props.id}`);
    expect(image).toBeInTheDocument();

    const description = screen.getByText(props.description);
    expect(description).toBeInTheDocument();

    const price = screen.getByText(`$ ${formatNumber(props.price)}`);
    expect(price).toBeInTheDocument();
    
    const conditionQuantity = screen.getByText(`${itemCondition(props.condition)} - ${props.soldQuantity} vendidos`);
    expect(conditionQuantity).toBeInTheDocument();

    const title = screen.getByText(props.title);
    expect(title).toBeInTheDocument();

    const buyButton = screen.getByText('Comprar');
    expect(buyButton).toBeInTheDocument();
  });

  it('should match the snapshot', () => {
    const { asFragment } = render(<ItemDetail {...props}/>);
    expect(asFragment()).toMatchSnapshot();
  });
});
