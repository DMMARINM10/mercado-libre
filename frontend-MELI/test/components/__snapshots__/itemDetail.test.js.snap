// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ItemDetail component should match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="item-detail-left"
  >
    <img
      alt="picture-item-1"
      height="680"
      src="http://random-url.jpg"
      width="680"
    />
    <div
      class="item-description-container"
    >
      <h2>
        Descripci√≥n del producto
      </h2>
      <p>
        Test description
      </p>
    </div>
  </div>
  <div
    class="item-detail-right"
  >
    <p
      class="item-detail-condition"
    >
      Nuevo - 10 vendidos
    </p>
    <div
      class="item-detail-metadata"
    >
      <h1>
        Test title
      </h1>
      <p>
        $ 1.000
        <span>
          10
        </span>
      </p>
      <button>
        Comprar
      </button>
    </div>
  </div>
</DocumentFragment>
`;
